### PR TITLE
Add ability to customize gas and gasPrice on tx confirmation screen

### LINKED ISSUE #765: Allow custom gas limits in send view

### PR DESCRIPTION
**Pull request by danfinlay**
Fixes #765

### CODE REVIEW

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -276,8 +275,9 @@ module.exports = class MetamaskController extends EventEmitter {
       exportAccount:             nodeify(keyringController.exportAccount).bind(keyringController),
 
       // txManager
-      approveTransaction:    txManager.approveTransaction.bind(txManager),
-      cancelTransaction:     txManager.cancelTransaction.bind(txManager),
+      approveTransaction:          txManager.approveTransaction.bind(txManager),
+      cancelTransaction:           txManager.cancelTransaction.bind(txManager),
+      updateAndApproveTransaction: this.updateAndApproveTx.bind(this),
```
Comment (flyswatter) [inline app/scripts/metamask-controller.js]: This is the big new action, it takes a `txMeta` object, updates it & signs it.

**Comment by kumavis** [comment]
such collab

File: app/scripts/lib/id-management.js [hunk 2]
```diff
@@ -25,13 +24,9 @@ function IdManagement (opts) {
...
   this.signTx = function (txParams) {
-    //  calculate gas with custom gas multiplier
-    var gasMultiplier = this.configManager.getGasMultiplier() || 1
-    var gasPrice = new BN(ethUtil.stripHexPrefix(txParams.gasPrice), 16)
-    gasPrice = gasPrice.mul(new BN(gasMultiplier * 100, 10)).div(new BN(100, 10))
-    txParams.gasPrice = ethUtil.intToHex(gasPrice.toNumber())
-    // normalize values
 
+    // normalize values
+    txParams.gasPrice = ethUtil.intToHex(txParams.gasPrice)
```
Comment (kumavis) [inline app/scripts/lib/id-management.js]: is this a safe coercion if `txParams.gasPrice` is already hex?

File: app/scripts/lib/config-manager.js [hunk 3]
```diff
@@ -228,18 +228,6 @@ ConfigManager.prototype._emitUpdates = function (state) {
...
-  return data.gasMultiplier
-}
-
-ConfigManager.prototype.setGasMultiplier = function (gasMultiplier) {
-  var data = this.getData()
-
-  data.gasMultiplier = gasMultiplier
-  this.setData(data)
-}
-
```
Comment (kumavis) [inline app/scripts/lib/config-manager.js]: ❤️

File: app/scripts/lib/id-management.js (hunk 2 above)
Comment (flyswatter) [inline app/scripts/lib/id-management.js]: This is a legacy file that I only updated to remove the gasMultiplier from.  We don't actually sign txs from it anymore.

File: ui/app/components/pending-tx-details.js [hunk 4]
```diff
@@ -191,6 +257,71 @@ PTXP.miniAccountPanelForRecipient = function () {
...
+    gas,
+    gasPrice,
+  })
+  const resultTxMeta = extend(txData, {
+    txParams: resultTx,
+  })
+  log.debug(`UI has computed tx params ${JSON.stringify(resultTx)}`)
+  return resultTxMeta
+}
+
```
Comment (kumavis) [inline ui/app/components/pending-tx-details.js]: this is way too much compute to be in a view. this should be in txManager.

File: ui/app/conf-tx.js [hunk 5]
```diff
@@ -48,7 +49,7 @@ ConfirmTxScreen.prototype.render = function () {
   var isNotification = isPopupOrNotification() === 'notification'
 
   log.info(`rendering a combined ${unconfTxList.length} unconf msg & txs`)
-  if (unconfTxList.length === 0) return null
+  if (unconfTxList.length === 0) return h(Loading)
```
Comment (kumavis) [inline ui/app/conf-tx.js]: 😸

File: ui/app/components/pending-tx-details.js (hunk 4 above)
Comment (flyswatter) [inline ui/app/components/pending-tx-details.js]: This is what it takes to live-update projected prices.

I'm ok with emitting the updated values to the background, but there's a big UX gain in the live-updating estimates that this provides.

File: ui/app/conf-tx.js (hunk 5 above)
Comment (flyswatter) [inline ui/app/conf-tx.js]: This doesn't seem to work right, though.  Loading is slow, and this view is often uncomfortably empty.  The reasons though, are deeply tangled.

File: ui/app/reducers/app.js [hunk 6]
```diff
@@ -291,7 +291,7 @@ function reduceApp (state, action) {
     case actions.SHOW_CONF_TX_PAGE:
       return extend(appState, {
         currentView: {
-          name: pendingTxs ? 'confTx' : 'account-detail',
+          name: 'confTx',
```
Comment (kumavis) [inline ui/app/reducers/app.js]: 👍

**Review by kumavis — APPROVED** [review]

File: ui/app/components/pending-tx-details.js (hunk 4 above)
Comment (kumavis) [inline ui/app/components/pending-tx-details.js]: > there's a big UX gain in the live-updating estimates that this provides

this seems like a safe assumption but curious if you tested it

Comment (flyswatter) [inline ui/app/components/pending-tx-details.js]: Of course.  It's really nice.  I encourage you to pull it down & try.

Comment (kumavis) [inline ui/app/components/pending-tx-details.js]: I mean, did you test if there was noticeable latency from doing it in the background

Comment (flyswatter) [inline ui/app/components/pending-tx-details.js]: Yeah, an early commit did it in the background, and it was laggy, and unresponsive, and there were also issues about the background state then overwriting the UI components, interrupting typing.

**Comment by kumavis** [comment]
LGTM

**Comment by kumavis** [comment]
>  for a parent to call a child's function.

yeah i think this is architecturally wrong, even if the framework supports it

**Comment by flyswatter** [comment]
I agree, it breaks the unidirectional data/action flow.  The alternative would be that the `details` should be part of the same component as this, since their concerns are so coupled.

**Comment by kumavis** [comment]
yeah we had them separate for the html popup read-only view
but doesnt really make sense

**Comment by flyswatter** [comment]
What's the html popup read-only view?

**Comment by flyswatter** [comment]
Oooh!  The old notification one!

**Comment by flyswatter** [comment]
Yeah, so combining these into one component would be a worthy task.

**Comment by flyswatter** [comment]
Should we do that before deploying this, or open an issue, you think?

**Comment by kumavis** [comment]
> open an issue

**Pull request merged**